Harvest restricted

cmeessen opened the pull request:
Fixes #312 

A second approach at issue #312 to avoid the growth of the harvester in certain directions. The optional parameter `restrict` now controls the general propagation behaviour of the harvester and disables growth in the given directions. By default `restrict` is `None`.

For the test I would suggest to plant one seed in a small mesh, restrict growth in five directions and then check the mesh for occupied cells after harvester finished.
### Checklist:
- [x] Make tests for new code (at least 80% coverage)
- [x] Create/update docstrings
- [x] Include relevant equations and citations in docstrings
- [x] Docstrings follow the style conventions
- [x] Code follows PEP8 style conventions
- [x] Code and docs have been spellchecked
- [x] Include new dependencies in `doc/install.rst`, `requirements.txt`, `environment.yml`, `ci/requirements-conda.txt` and `ci/requirements-pip.txt`.
- [x] Documentation builds properly (run `cd doc; make` locally)
- [x] Changelog entry (leave for last to avoid conflicts)
- [x] Firt-time contributor? Add yourself to `doc/contributors.rst` (leave for last to avoid conflicts)

leouieda commented on fatiando/gravmag/harvester.py:
```diff
@@ -264,7 +265,10 @@ def harvest(data, seeds, mesh, compactness, threshold, report=False):
+    * restrict : list of str
+        Restricts seed growth in given directions. Possible directions are 
+        ``'above'``, ``'below'``, ``'front'``, ``'back'``, ``'left'`` and
+        ``'right'``.
```
Please add something like: 

    You can pass in multiple directions as a list. Default is ``None`` for unrestricted growth.

leouieda commented on fatiando/gravmag/harvester.py:
```diff
@@ -526,34 +533,42 @@ def _is_neighbor(index, props, neighborhood):
 
 
-def _neighbor_indexes(n, mesh):
+def _neighbor_indexes(n, mesh, restrict=None):
```
The default values can be used only in the top-level function (`harvest` in this case) and it will be passed along to the other functions. If you put it everywhere, you'll have to remember to change all of them if you want to change the default.

leouieda commented:
@cmeessen this is great, thanks! 

I like your test idea. There is a simple test for the harvester code in the docstring of `harvest` is I'm not mistaken. You might be able to modify it for your purposes but please put it in a new file `fatiando/gravmag/tests/test_harvester.py`. Be sure to make a test that would fail if you don't restrict the growth (i.e., the result would grow out of the bounds you want if you don't restrict it). This can even be part of your test case. Testing for failure is also important as it makes sure the code does what you expect.

leouieda commented:
Also, the build is failing because of a PEP8 inconsistency. You can use `make pep8` from the main `Makefile` to check if your code complies. This one is just a trailing whitespace problem, nothing major.

cmeessen commented:
@leouieda thanks for taking the time looking into the code. I did indeed forget to `make pep8` before I committed, sorry for that. The changes you mentioned are now implemented into the new commit.

leouieda commented on fatiando/gravmag/harvester.py:
```diff
@@ -266,9 +266,10 @@ def harvest(data, seeds, mesh, compactness, threshold, report=False,
         ``'above'``, ``'below'``, ``'front'``, ``'back'``, ``'left'`` and
-        ``'right'``.
+        ``'right'``. You can pass in multiple directions as a list, e.g.
+        ``['above', 'front']``. Default is ``None`` for unrestricted growth.
```
This is perfect, thanks!

leouieda commented:
@cmeessen no worries about the `make pep8`.  I forget to run it all the time. That's why TravisCI checks this every time automatically.

Also, feel free to make as many commits as you want for the PRs. When we merge the PR, all  your commits will be collapsed into a single commit, so intermediate changes are squashed. This gives you more freedom to try things and revert them later without worrying about polluting the version history  of the repository.

leouieda commented:
Looks like now what we  need are a few tests. If you could also do a gallery example, that would be awesome.

Notice that the TravisCI build passed but coveralls is complaining that this PR decreases test coverage.

cmeessen commented:
I just finished the test for the restricted growth. A 3x3x3 `PrismMesh` is used to test the six possible restrictive cases by adding prisms to the center-prism in the direction of growth restriction. The harvester is then run twice, with and without restriction, using the center prism as initial seed. Finally, the resulting prism meshes are compared with the reference mesh. Again I think there is plenty of space for improvement though. At the moment the test only regards one restriction direction, but could be expanded to multiple directions.

I got a bit confused regarding the neighbouring indices: The `left` and `right` neighbours seem to have the indices 16 and 10. From this follows that the observer is facing south referring to left and right? I also suggest to use N/S/E/W instead, which should make it more intuitive to use.

leouieda commented:
@cmeessen perfect! The tests look great and cover all new code you added. We can worry about new tests later on. For now, this PR has been delayed too much already (sorry about that, it's my fault not yours).

I completely agree with renaming `front, back, left, right` to `north, south, east, west`. It makes much more sense avoids a lot of confusion. Just be aware that in Fatiando x points North and y points East, so for a prism `x1 = S, x2 = N, y1 = W, y2 = E`.

As soon as you finish this, add yourself to `doc/contributors.rst` and create a new entry in `doc/changelog.rst` (follow the style of the most recent ones). Let me know if you need help with this but don't worry too much about it. 

Let me know and I'll merge this. Thanks again for putting in the effort to implement this! Welcome to the team!

cmeessen commented:
@leouieda Thank you! I changed the names of the restriction directions, added a changelog entry and myself to the contributors list. Unfortunately couldn't make it before the release of 0.5, but looking forward to seeing this PR merged to fatiando. I'm glad to be able to contribute!

leouieda commented:
@cmeessen this is wonderful, I'm merging it in. No worries about 0.5, I want to have 0.6 not too long in the future. Thank you so much for doing this! I'll send you an invite to the Fatiando github organization. If you accept, you can set your status as private (no one will see that you're a member) or public (it shows in your github profile).

Also, I had some stickers made from the new logo and I'll get every contributors address early next year to send some as a way of saying thanks.

Either way, welcome to the team!
